src/hotspot/share/opto/vectornode.cpp

```diff
@@ -1131,42 +1141,66 @@ MacroLogicVNode* MacroLogicVNode::make(PhaseGVN& gvn, Node* in1, Node* in2, Node
 
 Node* VectorNode::degenerate_vector_rotate(Node* src, Node* cnt, bool is_rotate_left,
                                            int vlen, BasicType bt, PhaseGVN* phase) {
-  assert(bt == T_INT || bt == T_LONG, "sanity");
+  assert(is_integral_type(bt), "sanity");
   const TypeVect* vt = TypeVect::make(bt, vlen);
 
-  int shift_mask = (bt == T_INT) ? 0x1F : 0x3F;
-  int shiftLOpc = (bt == T_INT) ? Op_LShiftI : Op_LShiftL;
-  int shiftROpc = (bt == T_INT) ? Op_URShiftI: Op_URShiftL;
+  int shift_mask = (type2aelembytes(bt) * 8) - 1;
+  int shiftLOpc = (bt == T_LONG) ? Op_LShiftL : Op_LShiftI;
+  auto urshiftopc = [=]() {
+    switch(bt) {
+      case T_INT: return Op_URShiftI;
+      case T_LONG: return Op_URShiftL;
+      case T_BYTE: return Op_URShiftB;
+      case T_SHORT: return Op_URShiftS;
+      default: return (Opcodes)0;
+    }
+  };
+  int shiftROpc = urshiftopc();
 
   // Compute shift values for right rotation and
   // later swap them in case of left rotation.
   Node* shiftRCnt = NULL;
   Node* shiftLCnt = NULL;
-  if (cnt->is_Con() && cnt->bottom_type()->isa_int()) {
-    // Constant shift case.
-    int shift = cnt->get_int() & shift_mask;
+  const TypeInt* cnt_type = cnt->bottom_type()->isa_int();
+  bool is_binary_vector_op = false;
+  if (cnt_type && cnt_type->is_con()) {
+    // Constant shift.
+    int shift = cnt_type->get_con() & shift_mask;
     shiftRCnt = phase->intcon(shift);
     shiftLCnt = phase->intcon(shift_mask + 1 - shift);
-  } else {
-    // Variable shift case.
+  } else if (VectorNode::is_invariant_vector(cnt)) {
+    // Scalar variable shift, handle replicates generated by auto vectorizer.
     assert(VectorNode::is_invariant_vector(cnt), "Broadcast expected");
     cnt = cnt->in(1);
     if (bt == T_LONG) {
       // Shift count vector for Rotate vector has long elements too.
       assert(cnt->Opcode() == Op_ConvI2L, "ConvI2L expected");
       cnt = cnt->in(1);
     }
-    shiftRCnt = phase->transform(new AndINode(cnt, phase->intcon(shift_mask)));
+    shiftRCnt = cnt;
```

Comment:
Why do we remove the And with mask here?

Reply:
And'ing with shift_mask is already done on Java API side implementation before making a call to intrinsic rountine.

Reply:
This path is also taken from non vector api path, wont masking be needed there?

Reply:
@jatin-bhateja  This question is still pending.

Reply:
Other than VectorAPI , SLP also infers vector rotates where shift is either present in a vector or is a constant value. Also it will be illegal to perform any wrap around for shifts.

Reply:
@sviswa7, SLP flow will either have a constant 8bit shift value or a variable shift present in vector. So non constant scalar case will not be hit through this route.

Reply:
It would be better comment here, since the correctness relay on some others.

Reply:
@theRealELiu , @svisw7 , comment already exist in code,  I guess I mentioned incorrectly earlier on this thread, rectified my comments.